File: 1070/c.py
from collections import defaultdict
from operator import itemgetter


def find_index(list_to_check, index):
    left, right = 0, len(list_to_check) - 1
    while True:
        if left == right:
            return left
        mid = left + ((right - left) // 2)
        if list_to_check[mid] <= index <= list_to_check[mid]:
            return mid
        if list_to_check[mid] > index:
            right = mid
        else:
            left = mid+1


n, k, m = map(int, input().split())


tarif_plans = []
split_dots = set()
for i in range(m):
    l, r, c, p = map(int, input().split())
    tarif_plans.append((l, r+1, c, p))
    split_dots.add(l)
    split_dots.add(r+1)
tarif_plans.sort(key=itemgetter(3))

split_dots = sorted(split_dots)

date_ranges_tarif_plans = defaultdict(list)
for l, r, c, p in tarif_plans:
    lef = find_index(split_dots, l)
    rig = find_index(split_dots, r)
    for index, point in enumerate(split_dots[lef+1:rig+1], lef+1):
        date_ranges_tarif_plans[(split_dots[index-1], point)].append((c, p))

total_price = 0
for (s,e), ac in date_ranges_tarif_plans.items():
    cores = k
    days_diff = e-s
    for av_core, price in ac:
        if av_core > cores:
            total_price += days_diff * cores * price
            break
        else:
            total_price += days_diff * av_core * price
            cores -= av_core

print(total_price)

File: 1070/c.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

type Arrays [][4]int

func (a Arrays) Len() int           { return len(a) }
func (a Arrays) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
func (a Arrays) Less(i, j int) bool { return a[i][3] < a[j][3] }

func main() {
	in := bufio.NewReader(os.Stdin)
	line, _ := in.ReadString('\n')
	strs := strings.Split(strings.TrimSpace(line[:len(line)-1]), " ")
	base := [3]int{}
	for i, str := range strs {
		base[i], _ = strconv.Atoi(str)
	}
	tarifs := make(Arrays, base[2])
	dotsToSplitMap := make(map[int]bool)
	for i := 0; i < base[2]; i++ {
		line, _ := in.ReadString('\n')
		strs := strings.Split(strings.TrimSpace(line[:len(line)-1]), " ")
		for j, str := range strs {
			tarifs[i][j], _ = strconv.Atoi(str)
		}
		tarifs[i][1] += 1
		dotsToSplitMap[tarifs[i][0]] = true
		dotsToSplitMap[tarifs[i][1]] = true
	}
	sort.Sort(tarifs)
	dotsToSplit := make([]int, 0, len(dotsToSplitMap))
	for k := range dotsToSplitMap {
		dotsToSplit = append(dotsToSplit, k)
	}
	sort.Ints(dotsToSplit)
	dateRangesTarifPlans := make(map[[2]int][][2]int)
	for _, tarif := range tarifs {
		lef := findIndex(dotsToSplit, tarif[0])
		rig := findIndex(dotsToSplit, tarif[1])
		for i := lef + 1; i < rig+1; i++ {
			k := [2]int{dotsToSplit[i-1], dotsToSplit[i]}
			dateRangesTarifPlans[k] = append(dateRangesTarifPlans[k], [2]int{tarif[2], tarif[3]})
		}
	}
	totalPrice := 0
	for k := range dateRangesTarifPlans {
		cores := base[1]
		daysDiff := k[1] - k[0]
		for _, v := range dateRangesTarifPlans[k] {
			if v[0] > cores {
				totalPrice += daysDiff * cores * v[1]
				break
			} else {
				totalPrice += daysDiff * v[0] * v[1]
				cores -= v[0]
			}
		}
	}
	fmt.Println(totalPrice)
}

func findIndex(listToCheck []int, index int) int {
	left := 0
	right := len(listToCheck) - 1
	for {
		if left == right {
			return left
		}
		mid := left + ((right - left) / 2)
		if listToCheck[mid] <= index && index <= listToCheck[mid] {
			return mid
		}
		if listToCheck[mid] > index {
			right = mid
		} else {
			left = mid + 1
		}
	}
}
